Kernel/StackWriter.h

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright (c) 2022, Kim Nilsson <kim@wayoftao.net>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include <AK/StdLibExtraDetails.h>
+#include <AK/Types.h>
+
+namespace Kernel {
+
+class StackWriter {
+public:
+    explicit StackWriter(FlatPtr top)
+        : m_top { top }
+    {
+    }
+
+    // Decrement the stack top, reinterpret the intermediate bytes as an (uninitialized) object of type T, and return a reference to it.
+    // FIXME: This is probably technically UB, but works by tradition. See C++23 std::start_lifetime_as.
+    template<typename T>
+    [[nodiscard]] T& emplace(size_t offset = 0)
+    {
+        // Offsetting only makes sense if it is less than the size
+        // of the object we are trying to emplace.
+        VERIFY(offset < sizeof(T));
+
+        m_top -= (sizeof(T) - offset);
+
+        // We probably never want to use an unaligned address.
+        VERIFY(m_top % alignof(T) == 0);
+
+        return *reinterpret_cast<T*>(m_top);
```

Comment:
Use `bit_cast` here, also, as you so helpfully noted, this will need to go through `__builtin_launder` to be considered a valid object.

Reply:
How would bit_cast help in this situation? It returns an (efficient) copy of the referenced memory, whereas we would like to obtain a mutable reference.

Reply:
bit_cast would replace the reinterpret_cast, you'd effectively need `*__builtin_launder(bit_cast<T*>(m_top))`

Reply:
> `*__builtin_launder(bit_cast<T*>(m_top))`

Replacing reinterpret_cast with bit_cast here seems incorrect. reinterpret_cast is exactly what you want when converting between pointers and their implementation-defined integral representations. We won't be more defined by explicitly involving object representations of pointers contra integers (in-fact we may even be less defined if such a thing is possible).

As for launder, isn't it intended for the case where you have constructed an object A of type T at (pointer x) and then constructed another object B, also of type T, _at the same location_ with something like placement new, finally attempting to access it through (pointer x) which was derived from object A?

In essence the problem here has to do with lifetimes. We are commiting the sin of UB by assuming that there is an object of type T at the storage location given by m_top, which there isn't since we never constructed one there to begin with. We are actually pretty sure that there isn't one there. Indeed, thinking more about it now it seems the only valid way forward is to use placement new, which brings with it a can of worms in the form of the i686 RegisterState struct management.

In that particular piece of code we need to, depending on if we're returning to userspace or not, write either a full RegisterState struct, or omit the last two registers (userspace_ss, userspace_esp). The current solution is to offset the write pointer by the size of the registers in question and YOLO-cast it into a RegisterState struct, and then avoiding accessing the two fields, whose storage locations would in practice be somewhere at/above the kernel stack top.

So with placement new, how do we conditionally avoid writing the two registers? Split the types into something like RegisterState and PartialRegisterState? Well now we run into the problem that TrapFrame contains a pointer to the start of the RegisterState section, which is used to access some field that is common to both structs Effectively we are performing type erasure and would like to recover the original type again on the other side.

That's where I am right now, so what do you think?

Reply:
> Replacing reinterpret_cast with bit_cast here seems incorrect. reinterpret_cast is exactly what you want when converting between pointers and their implementation-defined integral representations

Not sure what you mean, this bit_cast is effectively the more restricted version of reinterpret_cast; `reinterpret_cast<T*>(FlatPtr)` just copies the bits into a pointer type (effectively), and so does `bit_cast<T*>(FlatPtr)` without looking like some reinterpret_cast fuckery is going on. this isn't about correctness, it's just about picking the least threatening cast :P 

Sidenote about pointers that aren't directly integers: we don't have those, half the universe would blow up if you tried to run Serenity code on such an architecture anyway.


> In essence the problem here has to do with lifetimes. We are commiting the sin of UB by assuming that there is an object of type T at the storage location given by m_top, which there isn't since we never constructed one there to begin with.

Yep, my assumption with this function was that the caller promises us to call the ctor (via placement new, or through some dark magic), hence the launder.

it seems like this thing is not going to conform to the "normal" lifetime model, so perhaps we could:
- static_assert that the type being given to this function is a POD type (so no UB if you skip the ctor and try to access the members)
- keep your impl, but slap a nice NOTE above the callsite.
- skip the launder as well as the lifetime note (no lifetime exists anyway)
- pray that the c++ gods don't smite us for our sins